### DISCUSSION TITLE
dynamic menu icon in a custom module

### DISCUSSION THREAD

**Discussion by mark-james** [directus/directus#16095] [Archive]
Hi,
Is it possible to change the icon used in a custom module dynamically?

I have a module that allows users to switch businesses to view different data sets. (A variation on a multi-tenant setup) The problem is the user can't tell what business they are currently active in without going back to the module to check which one is selected.

If the menu icon could be updated based on the business selected, the user could see the active business at a glance from anywhere in the app.

Thanks
Mark

**Comment by azrikahar** [comment]
There's no dynamic icon logic for such use case, at least as of now. However one potential way to achieve this would be to do this logic in the read hook for settings (when the app is being loaded, settings is read to populate the store, which includes module bars).

Here's a quick POC I've tested:

```js
import { defineHook } from '@directus/extensions-sdk';

export default defineHook(({ filter }, context) => {
	filter('settings.read', (input, meta, { accountability }) => {
		const moduleBars = input[0]?.module_bar;

		// if settings being queried doesn't contain `module_bar` (such as only translation_string being fetched), return early as we don't need to process it.
		if (!moduleBars) return input;

		// basic check if the current role is not the role we want to show different icon
		// if doesn't match, return it early as we don't need to process it.
		// you can also further elaborate this logic using RolesService instead of hardcoded ID like this.
		if (accountability?.role !== '4888bd03-ed1e-4ffa-bd56-8bcf5d8e5494') return input;

		input[0].module_bar = moduleBars.map((moduleBar) => {
			// override the original icon for our module based on name
			if (moduleBar?.name === 'Custom Module') {
				return { ...moduleBar, icon: 'work' };
			}

			return moduleBar;
		});

		return input;
	});
});
```

and here's the result:

https://user-images.githubusercontent.com/42867097/197124053-608ea2b5-adac-4750-ae14-d7d9f817af76.mp4

We can see originally the custom module has the icon `domain`, but when the role `4888bd03-ed1e-4ffa-bd56-8bcf5d8e5494` in my example code fetches the module bars, they are seeing the icon `work` instead.

I've added comments on how it can be further "enhanced", but it's intentionally simple for example purpose as it can depend on a case by case basis. For example, if you are using a custom field in `directus_users` for your multi-tenant setup, you'll then need to use UsersService to read said custom field to perform the `if` (or `switch`) statement to override the icon 👍

**Reply by mark-james** [reply]
Legend!!

Thanks for such a detailed reply. I'll give it a shot.
